[Doppel] Converting Doppel Alerts to STIX 2.1 Observables

siddharths1-metron opened the pull request:
<!--
Thank you very much for your pull request to the OpenCTI project! We as a community driven project depend on support and contributions like this!

Thus already a BIG THANK YOU upfront to you for choosing to help with your PR.
-->

### Proposed changes

* This new version of Doppel OpenCTI connector converts doppel alerts to STIX 2.1 Observables.
* Converts the Observables into Indicators if the alert is actioned/taken_down.
* Revokes the Indicator if alert is mark as unresolved from actioned/taken_down state.

### Related issues

* #4385

### Checklist

<!--
Please submit the source code in a way, where you could honestly say `This code is finished`.
If you feel that there are possibilities for improving the code quality, please do so.
By doing this, you are actively helping us to improve the quality of the entire OpenCTI project.
-->

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion)
- [x] Where necessary I refactored code to improve the overall quality

<!-- For completed items, change [ ] to [x]. -->
<!-- To sign commits with a GPG key, you can refer to https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits. -->

### Further comments

<!-- If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc... -->

Ninoxe commented:
Hi @siddharths1-metron 

Thank you for your PR.

There are a lot of files that have nothing to do with Doppel. Can you modify it so that only Doppel files are included in the PR?

siddharths1-metron commented:
Hello @Ninoxe any update on this? I did run the cmd as suggested. Thanks!

Ninoxe commented on external-import/doppel/src/doppel/converter_to_stix.py:
```diff
@@ -56,102 +78,632 @@ def _create_tlp_marking(level: str) -> Stix2MarkingDefinition:
+        # If not found and we have domain/IP, search by pattern
+        if (not indicators or len(indicators) == 0) and (domain_name or ip_address):
+            search_value = domain_name or ip_address
+            self.helper.log_info(
```
`self.helper.log_` is deprecated. As in the rest of the code, it is recommended to use `self.helper.connector_loger.error/info/warning`.

Furthermore, it is best to avoid using f strings in log messages. Prefer to pass arguments to the dictionary like this:
```
self.helper.connector_logger.error(
    "my message",
    {"error": e, "alert": alert, etc.}
)
```

Ninoxe commented on external-import/doppel/src/doppel/converter_to_stix.py:
```diff
@@ -56,102 +78,632 @@ def _create_tlp_marking(level: str) -> Stix2MarkingDefinition:
+            note_refs.append(primary_observable_id)
+
+        if note_refs:
+            reversion_note = Note(
```
The creation of a Stix object is done like this:
```
StixName(
    id=PyctiName.generate_id(…),
)
```
You need to import Notes from stix2 as well as from pycti, such as:
```
from pycti import Note as PyctiNote
[...]
reversion_note = Note(
    id=PyctiNote.generate_id(
        content=note_refs,
        created=modified,
    ),
[...]
```

The same applies to Indicator and Relationship.

siddharths1-metron replied:
Fixed, Thank you.

siddharths1-metron commented on external-import/doppel/src/doppel/converter_to_stix.py:
```diff
@@ -1,11 +1,33 @@
+    IPv4Address,
+)
 from stix2 import MarkingDefinition as Stix2MarkingDefinition
+from stix2 import Note
```
We have made the change, Thank you.

siddharths1-metron commented on external-import/doppel/src/doppel/converter_to_stix.py:
```diff
@@ -56,102 +78,632 @@ def _create_tlp_marking(level: str) -> Stix2MarkingDefinition:
+        :return: StixCoreRelationship object
+        """
+        relationship = StixCoreRelationship(
+            id=StixCoreRelationship.generate_id(
```
We have fixed this, thanks.

maximerafaillac commented:
Hello @Bharatkk-metron and @siddharths1-metron,

Have you seen this? https://github.com/OpenCTI-Platform/opencti/blob/33061911fd8b2c1e0f98f82602fb07cae6c2aab4/client-python/pycti/utils/constants.py#L548

There is a PhoneNumber Observable in OpenCTI. It's a custom Observable, meaning not a STIX Observable.

I hope it helps

siddharths1-metron commented:
Hey @maximerafaillac thank you for your input!

We will try to use PhoneNumber Observable for telco alerts and get back to you.

siddharths1-metron commented:
Hello @maximerafaillac, 

We have implemented the PhoneNumber Observable for telco alerts and able to see Observable in the OpenCTI dashboard. 

We have faced some limitations - 

- We want to have labels and external references to the Custom Phone Number Observable. We have checked the implementation you shared, it takes three params - value, spec_version and tracking-number. We dont see any supportive parameter for labels and external references. Can you please guide us for the same? 
- Our connector converts observable to Indicators. It working perfectly for domain alerts; when we try to do same for non-domain alerts (Custom Phone Number Observable) - the indicator is not created. Is there is any other way to create an Indicator for Phone Number Observable?

Below are the implementation details for creating Indicators - 

```
from pycti import (
    Indicator as PyctiIndicator
)
.
.
.

indicator = Indicator(
            id=PyctiIndicator.generate_id(pattern),
            pattern=pattern,
            pattern_type="stix",
            spec_version=STIX_VERSION,
            name=name,
            description=build_description(alert),
            created=created_at,
            modified=modified,
            created_by_ref=self.author.id,
            object_marking_refs=[self.tlp_marking.id],
            labels=labels_flat or None,
            external_references=external_references if external_references else None,
            valid_from=created_at,
            custom_properties=custom_properties,
            allow_custom=True,
        )
```

Could you please help us with the above points? 

Thank you.

Ninoxe commented:
Last error
<img width="1049" height="23" alt="image" src="https://github.com/user-attachments/assets/5e4aa185-0f82-4711-978d-89639477f06f" />

Final command to validate everything to do in /connectors:
`isort external-import/doppel/src/doppel/converter_to_stix.py`

And let's cross our fingers that this is the last modification 🤞

siddharths1-metron commented:
Hello @Ninoxe, That was our initial plan but I am not able to see any such checkbox. Hence we shared access this way. 

Thank you for understanding!

siddharths1-metron commented:
Hello @Ninoxe, Thank you very much for the detailed steps!

I have pushed the changes and here is the screenshot for your reference. 

<img width="1658" height="207" alt="image" src="https://github.com/user-attachments/assets/86e94c47-a2cf-48b7-845c-997b7ecb9576" />
